### ISSUE TITLE
RFC: Semi-sync management

### ISSUE THREAD

**Issue by deepthi** [vitessio/vitess#5505]
### Problem Definition
Vitess allows you to optionally use semi-sync at the mysql level for data durability. The behavior is gated by the flag `-enable_semi_sync`. The current behavior of the flag is to set semi_sync to true on MASTER and REPLICA type tablets, but not on RDONLY or other tablet types.

When you have a multi-cell deployment where a cell corresponds to a failure domain, it is desirable to achieve cross-cell durability. In practice, cells correspond to either zones (AZ in AWS) or regions, with zones being the more popular choice.

Today this is achieved by ensuring that a master and it's semi_sync eligible replica are located in different cells. This is cumbersome to manage in the deployment infrastructure. In addition, it leads to complexity around failovers, whether planned or unplanned. Also, this is only practical when the desired number of semi-sync acks is set to 1 (rpl_semi_sync_master_wait_for_slave_count = 1).

### Proposed Solution
The proposal is to have vitess manage semi-sync settings on REPLICA and RDONLY tablets to achieve durability guarantees.

We will add the concept of a Semi-Sync policy to vitess. The policy will be global to the vitess deployment, stored in the global topo and replicated in the cell-local topo. Having no policy will be the same as setting `-enable_semi_sync` to false. This flag will be deprecated.

Properties:
- DurabilityLevel: Allowed values: Cell, CellsAlias, CellsAliasGroup. DurabilityLevel of Cell means that semi-sync replicas are allowed in the same cell as the master. CellsAlias means that semi-sync replicas are allowed in the same CellsAlias but not in the same cell, and so on.
- NumberOfSemiSyncs: Default to 1. This value will be set on the master during my.cnf generation, and this number of replicas will have semi-sync enabled in each cell/cellsAlias.
- PreferAvailability: Defaults to false. Setting this to true will allow vitess to set semi-sync to true on all replicas in "other" cells/cellsAliases rather than on a subset determined by NumberOfSemiSyncs.
- HandleExternalReparents: Defaults to true. In case of an external reparent via some other tool like orchestrator, we will still need to fixup semi_sync to respect the policy.

###Pre-requisites
This feature should be used at a CellsAlias / Group level only if there are at least 3 cells in the deployment.

###Implementation
Various vtctl commands will need to change to respect the semi-sync policy
- InitShard
- InitShardMaster
- PlannedReparentShard
- TabletExternallyReparented

The general approach is to change SetMaster to be cell-aware. In addition, replication_reporter will fixup semi_sync to respect the policy in order for this to work correctly with external reparents.

**Comment by aquarapid** [comment]
To be clear, DurabilityLevel will default to Cell?

**Comment by deepthi** [comment]
> To be clear, DurabilityLevel will default to Cell?

Yes. That is current behavior with -enable_semi_sync and we will keep that as default.

**Comment by demmer** [comment]
We (Slack) had a somewhat involved discussion about this exact issue the other day.
I have a couple comments on the proposal:

> DurabilityLevel: Allowed values: Cell, CellsAlias, CellsAliasGroup. DurabilityLevel of Cell means that semi-sync replicas are allowed in the same cell as the master. CellsAlias means that semi-sync replicas are allowed in the same CellsAlias but not in the same cell, and so on.

I feel like this is a non-intuitive way to describe the policy as I would interpret a DurabilityLevel of "Cell" to mean that writes are cross-cell. It would be clearer (to me) if the option was more clearly tied to semi-sync and specified the achieved durability level.

Furthermore the current proposal doesn't actually seem to have an option that covers today's default behavior where REPLICA and not RDONLY are eligible. How about:

SemiSyncEligibility:
   * None: no semi sync at all (default if `enable_semi_sync` is false)
   * AnyReplica: any REPLICA tablet is eligible (default if `enable_semi_sync` is true )
   * Any: any tablet is eligible
   * CrossCell: any tablet not in the same cell as the master is eligible
   * CrossCellAlias: any tablet not in the same cell alias as the master is eligible
   * CrossCellGroup: any tablet not in the same cell alias group as the master is eligible

Also there is a concern worth calling out relating to external reparents:

>  In addition, replication_reporter will fixup semi_sync to respect the policy in order for this to work correctly with external reparents.

Since this change happens asynchronously it introduces a window of time when there is a risk  of violating the policy
 
Suppose the policy is CrossCell and orchestrator reparents to make the master in the same cell as a replica, then that replica will send semi-sync acks until the replication reporter runs to stop it from doing semi-sync.

This introduces a small window where the durability policy is not enforced.

If (when?) Vitess implements native failover handling to replace the need for orchestrator, then this window of time would not exist because the policy would be enforced when the master promotion occurs, not after the fact.

**Comment by deepthi** [comment]
@demmer thank you for the feedback. You make some excellent points. I will incorporate them into the proposal.
